### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by wujingtao0730** [ultralytics/ultralytics#22791]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When I train YOLOE using three datasets according to the document, I end up with a RAM OOM error, and the memory usage increases with the number of training iterations. I searched for some blogs online, and some bloggers said that the memory leak was caused by the list in the dataloader. I tried setting workers=0, which solved the problem of memory leakage, but the training was too slow. Can this issue be addressed and fixed？

### Environment

Ultralytics 8.3.231 🚀 Python-3.9.18 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 10821MiB)
                                                       CUDA:1 (NVIDIA GeForce RTX 2080 Ti, 10823MiB)
                                                       CUDA:2 (NVIDIA GeForce RTX 2080 Ti, 10823MiB)
                                                       CUDA:3 (NVIDIA GeForce RTX 2080 Ti, 10823MiB)

### Minimal Reproducible Example

from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOESegTrainerFromScratch

data = dict(
    train=dict(
        yolo_data=["Objects365v1.yaml"],
        grounding_data=[
            dict(
                img_path="/opt/flickr/full_images",
                json_file="/opt/flickr/annotations/final_flickr_separateGT_train_segm.json",
            ),
            dict(
                img_path="/opt/mixed_grounding/gqa/images",
                json_file="/opt/mixed_grounding/annotations/final_mixed_train_no_coco_segm.json",
            ),
        ],
    ),
    val=dict(yolo_data=["lvis.yaml"]),
)

model = YOLOE("yoloe-11n-seg.yaml")
model.train(
    data=data,
    batch=64,
    epochs=30,
    close_mosaic=2,
    optimizer="AdamW",
    lr0=2e-3,
    warmup_bias_lr=0.0,
    weight_decay=0.025,
    momentum=0.9,
    workers=0,
    trainer=YOLOESegTrainerFromScratch,
    device="0,1,2,3",
)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wujingtao0730, thank you for your detailed report and for using Ultralytics 🚀! This is an automated triage response 🤖 — an Ultralytics engineer will follow up with you here soon for a deeper look.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many common questions may already be answered.

Since this is a 🐛 Bug Report related to training (potential memory leak / RAM OOM), your provided **Minimal Reproducible Example (MRE)** is very helpful. If possible, you can further help by confirming:
- Whether the issue reproduces with a smaller subset of your datasets
- Whether it appears only with YOLOE / `YOLOESegTrainerFromScratch` or also with other trainers/models

If this is or becomes a custom training ❓ question (for example, if you want advice on how to structure multi-dataset training or optimize workers/batch sizes), please provide as much information as possible, including:
- Dataset image examples (or structure descriptions)
- Full training logs
- Any custom changes to the code or configs
And verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/)  
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community 💬  

## Upgrade

Before an engineer investigates further, please ensure you have tested with the latest `ultralytics` package and dependencies to confirm the issue still occurs, as many bugs are fixed in recent releases.

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If you are able to reproduce the behavior in one of these verified environments (for example, with a smaller public dataset and your same YOLOE training script), that will make it much easier for engineers to confirm and debug any potential memory leak.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once again, thanks for the detailed environment info and MRE — this will help the team investigate your YOLOE training memory issue more efficiently 🙏

**Comment by wujingtao0730** [comment]
1. Currently, I am using Ultralytics 8.3.231 and only encountering this issue while training YOLOE's YOLOESegTrainerFromScratch。Perhaps the dataset is too large, causing a memory leak when reading the list. I have read some blogs that suggest using numpy.array can solve this problem, but due to the large amount of code in ultralytics, I did not attempt it
2. The dataset, I used objectis365v1, Flickr, mixed_ground, and lvis according to the official documentation, consistent with the official tutorial
3. Here is my training log, but it is difficult to reflect memory leaks in the training log. I can only see that there is not enough memory to end the training at the end
[yoloe-11n-seg-20251124-193147.log](https://github.com/user-attachments/files/23741011/yoloe-11n-seg-20251124-193147.log)

**Comment by glenn-jocher** [comment]
Thanks for the detailed repro and extra context — this does look like a YOLOE-specific host‑memory growth when using `YOLOESegTrainerFromScratch` with multiple datasets as in the [YOLOE training docs](https://docs.ultralytics.com/models/yoloe/).  

To help us pin this down, could you please:  
1) Run `yolo checks` (or `!yolo checks` in a notebook) and paste the full text output, plus confirm that there’s only one `ultralytics` installation in your environment.  
2) Share whether RAM still grows monotonically if you:  
   • keep the same script but set `workers=2` and `batch=16`, and  
   • train only on `Objects365` (remove `grounding_data`) for a short run.  
3) Confirm whether GPU memory reported by `nvidia-smi` stays roughly constant while system RAM grows, or if both increase.  

If you can also attach the output of this helper, it will give us more insight into your exact setup:  

```python
from ultralytics.utils.checks import collect_system_info
print(collect_system_info())
```  

Once we have those comparisons and system details, we can try to reproduce on our side and inspect the YOLOE dataloader path for any lists or caches that aren’t being released between iterations.

**Comment by wujingtao0730** [comment]
1. I reinstalled the latest ultralytics today， yolo checks result
Ultralytics 8.3.232 🚀 Python-3.9.18 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 10821MiB)
Setup complete ✅ (48 CPUs, 125.8 GB RAM, 4318.1/5543.9 GB disk)

OS                     Linux-5.15.0-139-generic-x86_64-with-glibc2.17
Environment            Docker
Python                 3.9.18
Install                pip
Path                   /root/anaconda3/envs/pytorch/lib/python3.9/site-packages/ultralytics
RAM                    125.76 GB
Disk                   4318.1/5543.9 GB
CPU                    Intel Xeon E5-2678 v3 2.50GHz
CPU count              48
GPU                    NVIDIA GeForce RTX 2080 Ti, 10821MiB
GPU count              4
CUDA                   11.7

numpy                  ✅ 1.26.2>=1.23.0
matplotlib             ✅ 3.8.2>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.1.0>=7.1.2
pyyaml                 ✅ 6.0.1>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.11.4>=1.4.1
torch                  ✅ 1.13.1+cu117>=1.8.0
torch                  ✅ 1.13.1+cu117!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.14.1+cu117>=0.9.0
psutil                 ✅ 5.9.6>=5.8.0
polars                 ✅ 1.34.0>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

2.  collect_system_info results
Ultralytics 8.3.232 🚀 Python-3.9.18 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 10821MiB)
Setup complete ✅ (48 CPUs, 125.8 GB RAM, 4318.1/5543.9 GB disk)

OS                     Linux-5.15.0-139-generic-x86_64-with-glibc2.17
Environment            Docker
Python                 3.9.18
Install                pip
Path                   /root/anaconda3/envs/pytorch/lib/python3.9/site-packages/ultralytics
RAM                    125.76 GB
Disk                   4318.1/5543.9 GB
CPU                    Intel Xeon E5-2678 v3 2.50GHz
CPU count              48
GPU                    NVIDIA GeForce RTX 2080 Ti, 10821MiB
GPU count              4
CUDA                   11.7

numpy                  ✅ 1.26.2>=1.23.0
matplotlib             ✅ 3.8.2>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.1.0>=7.1.2
pyyaml                 ✅ 6.0.1>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.11.4>=1.4.1
torch                  ✅ 1.13.1+cu117>=1.8.0
torch                  ✅ 1.13.1+cu117!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.14.1+cu117>=0.9.0
psutil                 ✅ 5.9.6>=5.8.0
polars                 ✅ 1.34.0>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18
{'OS': 'Linux-5.15.0-139-generic-x86_64-with-glibc2.17', 'Environment': 'Docker', 'Python': '3.9.18', 'Install': 'pip', 'Path': '/root/anaconda3/envs/pytorch/lib/python3.9/site-packages/ultralytics', 'RAM': '125.76 GB', 'Disk': '4318.1/5543.9 GB', 'CPU': 'Intel Xeon E5-2678 v3 2.50GHz', 'CPU count': 48, 'GPU': 'NVIDIA GeForce RTX 2080 Ti, 10821MiB', 'GPU count': 4, 'CUDA': '11.7', 'Package Info': {'numpy': '✅ 1.26.2>=1.23.0', 'matplotlib': '✅ 3.8.2>=3.3.0', 'opencv-python': '✅ 4.11.0.86>=4.6.0', 'pillow': '✅ 11.1.0>=7.1.2', 'pyyaml': '✅ 6.0.1>=5.3.1', 'requests': '✅ 2.32.3>=2.23.0', 'scipy': '✅ 1.11.4>=1.4.1', 'torch': '✅ 1.13.1+cu117!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.14.1+cu117>=0.9.0', 'psutil': '✅ 5.9.6>=5.8.0', 'polars': '✅ 1.34.0>=0.20.0', 'ultralytics-thop': '✅ 2.0.18>=2.0.18'}}

3. According to your request, I modified the training code，RAM Memory still increases with the number of iterations(the increase of memory consumption is relatively stable), the GPU memory will also increase during the training, but I'm not sure whether this is normal.
data = dict(
    train=dict(
        yolo_data=["./Objects365v1.yaml"],
        # grounding_data=[
        #     dict(
        #         img_path="/opt/flickr/full_images",
        #         json_file="/opt/flickr/annotations/final_flickr_separateGT_train_segm.json",
        #     ),
        #     dict(
        #         img_path="/opt/mixed_grounding/gqa/images",
        #         json_file="/opt/mixed_grounding/annotations/final_mixed_train_no_coco_segm.json",
        #     ),
        # ],
    ),
    val=dict(yolo_data=["./lvis.yaml"]),
)

model = YOLOE("yoloe-11n-seg.yaml")
model.train(
    data=data,
    batch=16,
    epochs=30,
    close_mosaic=2,
    optimizer="AdamW",
    lr0=2e-3,
    warmup_bias_lr=0.0,
    weight_decay=0.025,
    momentum=0.9,
    workers=2,
    trainer=YOLOESegTrainerFromScratch,
    device="0,1,2,3",
)

<img width="726" height="442" alt="Image" src="https://github.com/user-attachments/assets/6e07d244-47c5-4815-88cb-a8b982f8509b" />

<img width="722" height="429" alt="Image" src="https://github.com/user-attachments/assets/6d1f5802-9517-413e-a2f0-102fe59baf2d" />

**Comment by wujingtao0730** [comment]
When I set workers=0, there is no memory leak, but the training is slow

**Comment by glenn-jocher** [comment]
That’s very helpful confirmation – the fact that `workers=0` stops the growth strongly points to a YOLOE `YOLOESegTrainerFromScratch` multi‑worker dataloader issue rather than the core training loop, which already calls `_clear_memory()` each epoch to clear CUDA cache as shown in the [BaseTrainer reference](https://docs.ultralytics.com/reference/engine/trainer/#basetrainer). For now the safest workaround is to keep `workers` as low as you can tolerate (please also try `workers=1` with a reduced `batch`), and if possible test once in a fresh env with a newer PyTorch 2.x to see if the behaviour changes; we’ll treat this as a YOLOE bug on our side and investigate, then update this thread once we have a concrete fix.

**Comment by wujingtao0730** [comment]
Ultralytics 8.3.230 🚀 Python-3.11.14 torch-2.9.0+cu128 CUDA:0 (NVIDIA GeForce RTX 5090, 32109MiB)
Setup complete ✅ (56 CPUs, 125.8 GB RAM, 4975.1/7391.2 GB disk)

OS                     Linux-5.15.0-139-generic-x86_64-with-glibc2.35
Environment            Docker
Python                 3.11.14
Install                pip
Path                   /opt/conda/lib/python3.11/site-packages/ultralytics
RAM                    125.76 GB
Disk                   4975.1/7391.2 GB
CPU                    Intel Xeon CPU E5-2680 v4 @ 2.40GHz
CPU count              56
GPU                    NVIDIA GeForce RTX 5090, 32109MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.28.2>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.0+cu128>=1.8.0
torch                  ✅ 2.9.0+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.0+cu128>=0.9.0
psutil                 ✅ 7.1.0>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

Today I tried a new environment, as shown above: torch-2.9.0 + cu128, and I found that the memory leak issue disappeared. The previous memory leak problem might have been caused by torch 1.x. Perhaps adding PyTorch version information to the YOLOE training documentation could prevent others from encountering similar issues.

<img width="1126" height="85" alt="Image" src="https://github.com/user-attachments/assets/acc58a46-3246-4778-8e3c-c797b09673c5" />

**Comment by glenn-jocher** [comment]
Thanks for tracking this down and for testing with the newer stack — this does make it look like a PyTorch 1.13.x DataLoader/multiprocessing issue rather than a YOLOE‑specific leak; we’ll note this in the YOLOE training docs as a recommendation to use PyTorch 2.x (with a fallback of reducing `workers` on 1.x), and if you see similar RAM growth again on 2.x with this setup, please add a short log snippet or repro so we can revisit.
